Release GIL when doing the compute-intensive CRC computation

jonded94 opened the pull request:
Unfortunately, this library captures the Python GIL for the entirety of its computation. I chose this library because it's quite fast compared to others, but this is pretty worthless if it's forcing one to use multiprocessing for a realtively mundane task such as computing a hash of a given bytebuffer.

I think this should work. On a very simply test benchmark, I saw a sizeable performance improvement using this fork (~46s vs. ~8s using 16 threads on my 13th Gen Intel(R) Core(TM) i7-1370P):
```
import crc32c
import concurrent.futures
import io
import time


def calc_crc(inp: io.BytesIO) -> int:
    return crc32c.crc32c(inp.getbuffer())


N_DATA = 1_000_000_000
N_COUNT = 1000

data = io.BytesIO(b"0"*N_DATA)

ts = time.perf_counter()
res = [calc_crc(data) for _ in range(N_COUNT)]
te = time.perf_counter()
print(te-ts)  # 46.33s


with concurrent.futures.ThreadPoolExecutor(max_workers=16) as executor:
    ts = time.perf_counter()
    res_threaded = list(executor.map(calc_crc, [data for _ in range(N_COUNT)]))
    te = time.perf_counter()
    print(te-ts)  # 8.02s
    assert res == res_threaded
```

<!-- Generated by sourcery-ai[bot]: start summary -->

## Summary by Sourcery

Enhance the CRC computation by releasing the GIL during the compute-intensive operation, resulting in significant performance improvements in multi-threaded scenarios.

Enhancements:
- Release the Global Interpreter Lock (GIL) during the compute-intensive CRC computation to improve performance in multi-threaded environments.

<!-- Generated by sourcery-ai[bot]: end summary -->

jonded94 commented on _crc32c.c:
```diff
@@ -55,9 +55,11 @@ PyObject* crc32c_crc32c(PyObject *self, PyObject *args) {
 		return NULL;
 
 	bin_data = pbin.buf;
+	Py_BEGIN_ALLOW_THREADS
 	crc ^= 0xffffffff;
 	result = crc_fn(crc, bin_data, pbin.len);
 	result ^= 0xffffffff;
+	Py_END_ALLOW_THREADS
```
I actually benchmarked it against a ~1GB byte buffer and there releasing the GIL gave me an ~6x improvement of performance already on my developer machine.
On an even more multi-threaded compute server, the impact surely is even larger.

rtobar commented:
@jonded94 thanks for this contribution. It's been a long time since I wrote the package originally, and after so long I had assumed we were releasing the GIL already.

I agree with the change in principle; as you point out it unnecessary hinders multithreaded computation. Having said that, I'd like to have a look at the performance penalty as a function of I out buffer size. Surely there's a cost associated with releasing/acquiring the GIL, which might affect performance in single-threaded cases, specially with small buffer sizes. I seem to remember `binutils.crc32` does (or did) something similar.

I'll run some benchmarks locally and post results here. But again, that ks again for the contribution, and definitely a version of these changes will make it in.

rtobar commented:
@jonded94 thanks again for taking interest in this issue, and providing the patch.

I wrote the following benchmark to measure the overhead of releasing the GIL  as a function of input buffer size, both for the software and hardware modes.

```python
import time

import seaborn as sns
from pandas import DataFrame
from matplotlib import pyplot as plt

from crc32c import crc32c

def _run(buf, n_iter: int, release_gil: bool, sw_mode: bool) -> float:
    start = time.monotonic()
    [crc32c(buf, 0, release_gil, sw_mode) for _ in range(n_iter)]
    return time.monotonic() - start

def _run_with_bufsizes(n_iter: int) -> DataFrame:
    bufsizes = (2 ** n for n in range(1, 20))
    buffers = [b'0' * size for size in bufsizes]
    return DataFrame.from_records(
        [(len(buffer), _run(buffer, n_iter, False, sw_mode), _run(buffer, n_iter, True, sw_mode), sw_mode) for buffer in buffers for sw_mode in (False, True)],
        columns=("bufsize", "time_no_release", "time_release", "sw_mode")
    )


n_iter = 10000
df = _run_with_bufsizes(n_iter)
df["release_overhead"] = (df["time_release"] / df["time_no_release"]) - 1
sns.set_theme()
sns.catplot(df, x="bufsize", y="release_overhead", hue="sw_mode", kind="bar")
plt.show()
```

You'll note that the `crc32c` function used for this accepts two extra options to release/not-release the GIL, and force using the SW implementation or not. This is the diff on top of your changes:

```diff
--- a/_crc32c.c
+++ b/_crc32c.c
@@ -41,6 +41,8 @@ PyObject* crc32c_crc32c(PyObject *self, PyObject *args) {
        Py_buffer pbin;
        unsigned char *bin_data = NULL;
        uint32_t crc = 0U, result;
+       int release_gil = 0;
+       int sw_mode = 0;
 
        /* In python 3 we accept only bytes-like objects */
        const char *format =
@@ -49,17 +51,26 @@ PyObject* crc32c_crc32c(PyObject *self, PyObject *args) {
 #else
        "s*"
 #endif
-       "|I:crc32";
+       "|Ipp:crc32c";
 
-       if (!PyArg_ParseTuple(args, format, &pbin, &crc) )
+       if (!PyArg_ParseTuple(args, format, &pbin, &crc, &release_gil, &sw_mode) )
                return NULL;
 
+       crc_function the_crc_fn = (sw_mode ? _crc32c_sw_slicing_by_8 : crc_fn);
+
        bin_data = pbin.buf;
+       if (release_gil) {
        Py_BEGIN_ALLOW_THREADS
        crc ^= 0xffffffff;
-       result = crc_fn(crc, bin_data, pbin.len);
+       result = the_crc_fn(crc, bin_data, pbin.len);
        result ^= 0xffffffff;
        Py_END_ALLOW_THREADS
+       }
+       else {
+       crc ^= 0xffffffff;
+       result = the_crc_fn(crc, bin_data, pbin.len);
+       result ^= 0xffffffff;
+       }
 
        PyBuffer_Release(&pbin);
        return PyLong_FromUnsignedLong(result);
```

These are the results on my system (AMD Ryzen 7 5825U, CPython 3.12):

![benchmark](https://github.com/user-attachments/assets/cb3767e4-6cdc-435b-95ef-2d36446bda56)

As expected, the penalty is greater at smaller buffer sizes, and for the HW mode. Very roughly, and of course very specifically to my system, the penalty in HW mode decreases to ~2% at 32KB, and ~1% at 128KB. For SW those happen at somewhere around ~4KB and at ~8KB.

Also, as I remembered, the `binascii.crc32` function does indeed conditionally release the GIL, see https://github.com/python/cpython/blob/cc6839a1810290e483e5d5f0786d9b46c4294d47/Modules/binascii.c#L772-L798. They put the limit at 5 KB, which more or less matches the SW mode at somewhere the ~1/2%.

In summary, I think we should restrict the releasing of the GIL to a minimum buffer size.

I'm a bit torn on the actual value though: it should be small enough that it covers most user cases, but big enough that the overhead of releasing the GIL isn't too big. A buffer size that I think would be suitable would be similar to that used for file or socket reading operations, since the data fed into this package comes likely from those. These buffer sizes are usually in the single or double-digit KBs. And given the numbers I got in my system, I'm inclined to make the cut at 32 KB. @jonded94 would you want to put that change in?

If we also wanted more flexibility, we could offer a module-level function that altered this limit, or even a flag like in the diff above that can be used on a per-call basis.

jonded94 commented:
Thank you very much for your in-depth analysis, @rtobar ! Yes, it probably makes sense only releasing GIL after a certain threshold. I actually was unaware that it creates so much of an overhead.

**TODOs**:

- Implement dynamic GIL release behaviour, probably with this final function header: `crc32c(buf: typing.buffer, checksum: int = 0, release_gil: bool | None = None)`
  - My idea would be that `release_gil = None` should dynamically determine whether to release the GIL (i.e. after a 32KiB threshold for example). The user can specify on its own if he knows it better by giving a boolean explicitly.
- Make this package a PEP 561 (https://peps.python.org/pep-0561/) compliant package
  - Because of a missing stub files, `mypy` will error with this message: `error: Skipping analyzing "crc32c": module is installed, but missing library stubs or py.typed marker  [import-untyped]` 
  - Since there practically are only two functions in this package (with one being deprecated) I think it should be fine to include this in the PR? I thought of this because we're planning to slightly change the function header anyways, so proper type hints could be nice.

Will do this as soon as I find the time for it (probably this weekend or slightly later) :)

jonded94 commented:
I now made this package PEP 561 compliant and added typehint stub files. This seems to have worked, now even a `crc32c.crc32c(io.BytesIO(b"123").getbuffer())' properly typechecks with `mypy --strict`.

Also, I added a new keyword argument `release_gil`. This can be `bool | int`, and has this behaviour:
- (default) -1 (or less): Automatically decide whether to release GIL or not. Will do it starting from buffer sizes >= 32KiB
- 0: Never release GIL
- 1: Always release GIL

rtobar reviewed (commented):
Thank you very much @jonded94 for the further changes :)

Two things that I'd like to point out here:
 * The new GIL releasing behavior, and the new argument, need to be described in the documentation (i.e., the `README` file).
 * Similarly, let's add a new entry to the `CHANGELOG` file, under the Development section.
 * I'm happy to have annotations added, but let's leave that for a different PR to avoid mixing things together.

rtobar commented on test/test_crc32c.py:
```diff
@@ -80,6 +80,9 @@ class TestMisc(unittest.TestCase):
     def test_zero(self):
         self.assertEqual(0, crc32c.crc32c(b''))
 
+    def test_keyword(self):
+        self.assertEqual(10, crc32c.crc32c(b'', crc=10))
+
```
To be clear, the test I would envision would run crc32c with the different GIL behaviour, and check that calculated values are equal. It doesn't prove that the GIL was released or not, but it proves that we haven't broken things.

jonded94 replied:
Kept this test in as it *is* a change that now proper `kwargs` are allowed in general in this function.
Added a new function that "tests" the `gil_release_mode` by checking all possible modes for a small & a large (i.e. >32KiB) input.

rtobar replied:
OK, let's leave it in, the new test is really what I was after, I don't mind too much having this one in as well.

jonded94 commented on _crc32c.c:
```diff
@@ -37,10 +37,13 @@ static crc_function crc_fn;
 	uint32_t crc = 0U, result;
+	int release_gil = -1;
+
+	static char *kwlist[] = {"inp", "crc", "release_gil", NULL};
```
Changed the corresponding `kwargs`, thanks, I actually did not have a look at `binascii` & `zlib` for that.

jonded94 commented on _crc32c.c:
```diff
@@ -49,29 +52,37 @@ PyObject* crc32c_crc32c(PyObject *self, PyObject *args) {
+		Py_BEGIN_ALLOW_THREADS
+		crc ^= 0xffffffff;
+		result = crc_fn(crc, bin_data, pbin.len);
+		result ^= 0xffffffff;
```
Added a new `static inline int crc32c_inline` function.

jonded94 commented on _crc32c.c:
```diff
@@ -49,29 +52,37 @@ PyObject* crc32c_crc32c(PyObject *self, PyObject *args) {
-	crc ^= 0xffffffff;
-	result = crc_fn(crc, bin_data, pbin.len);
-	result ^= 0xffffffff;
+	if ((release_gil < 0 && pbin.len >= 32768) || release_gil >= 1) {
```
Added a macro called MIN_BUFSIZE_FOR_AUTOMATIC_RELEASE. Renamed `release_gil` to `gil_release_mode`.

Regarding the whole `int | bool` situation: I feel some of the interfaces of this package are a bit un-pythonic right now (i.e. setting the "sw mode" via an environment variable instead of a `kwarg`) and a `gil_release_mode` that maps magic integer values against different behaviour (which you would have to consult documentation for to know what it actually does) feels way more C-esque than pythonic to me.

jonded94 replied:
Added a macro called MIN_BUFSIZE_FOR_AUTOMATIC_RELEASE. Renamed release_gil to gil_release_mode.

Regarding the whole `int | bool` situation: I feel some of the interfaces of this package are a bit un-pythonic right now (i.e. setting the "sw mode" via an environment variable instead of a kwarg) and a `gil_release_mode` that maps magic integer values against different behaviour (which you would have to consult documentation for to know what it actually does) feels way more C-esque than pythonic to me. I don't have an opinion on this right now since if you actively want to change this value, you're probably doing some in-depth stuff anyways.

A `release_gil: bool | None = None` still feels way more self-descriptive to me since it describes pretty accurate what's it's doing just from the parameter name and its given typehints already. I did not implement it this way though since it requires some additional boilerplate on the C side, i.e. explicitly checking a `PyObject*` against `Py_False`, `Py_True`, `Py_None` (since `None` is falsy, so one can't simply check for booleaness).

rtobar replied:
Thanks for the further changes!

You are right that using an `int` is an overkill for a flag that chooses between three states. Ideally that would be an enum, but then again that's an even higher overkill given that this is all a C extension -- much easier to deal with a simple `int` that declare a python enum, handle it in C, etc...

Let's continue any further discussion about the type hinting in the other PR. I still think putting the `int | bool` is unnecessary complex (and incidentally the AI bot thing that commented above -- I'm not sure why it's coming up, it probably was enabled in our organisation -- also thought so).

The environment variable is partly fine I think. I do agree that it could be complemented with a kwarg in the `crc32c` function (very easy now that we support them! feel free to add a PR for that :wink:), but it's also a very easy toggle that can be used from any application without code changes.

jonded94 commented on _crc32c.c:
```diff
@@ -49,29 +52,37 @@ PyObject* crc32c_crc32c(PyObject *self, PyObject *args) {
-	                 "crc32c.crc32 will be eventually removed, use crc32c.crc32c instead",
-	                 1) == -1) {
+					 "crc32c.crc32 will be eventually removed, use crc32c.crc32c instead",
+					 1) == -1) {
```
These are two lines which *partly* were indented by spaces, ontop of the tab-based indentation that seems to be the case for every other line. I cleaned that up to have roughly the same indentation, but with tabs.

If you really don't want *any* formatting changes in this MR, please let me know. I figured fine since it's just two lines and the only place in this file where indentation is a bit messy, it's fine to do it right now as long as it is in a separate commit.

rtobar replied:
Yes, I'd rather not include any formatting changes if possible. I used to be strongly minded about tabs/spaces (indent with tabs, align with spaces), but with time I've accepted the fact that most people/editors/IDEs simply work better with only spaces. Still, I'd rather keep formatting differences away. Thanks!

jonded94 replied:
Removed the formatting. Never thought I'd have a discussion about tabs vs. spaces in the 21st century :smile:

jonded94 commented:
Thanks [rtobar](https://github.com/rtobar) for your in-depth review! :)

I hopefully addressed all of your concerns and pushed a bunch of new commits. Specifically, I also removed the adding of typehints from this PR and moved them into a separate one (https://github.com/ICRAR/crc32c/pull/49).

Please let me know if you see anything that still needs further refinement.

rtobar reviewed (commented):
First of all, thank you so much @jonded94 for the continuing effort and patience while going through this review. This is looking excellent! Just a few more minor comments. Could you also rebase on top of the latest master? If possible I'd also like to push just one or t commit instead of 11 so far.

jonded94 commented on _crc32c.c:
```diff
@@ -32,15 +32,28 @@
+	crc ^= 0xffffffff;
+	result = crc_fn(crc, bin_data, len);
+	result ^= 0xffffffff;
+	return result;
```
My editor (JetBrains family) *really* doesn't like mixed indentation (i.e. *both* tabs and spaces) .. :)

jonded94 commented:
Implemented the suggestions from your review and squashed the changes into one commit. Should be fine now? :smile:

rtobar commented:
Many thanks again @jonded94 for the effort and patience! I'll wait for CI to check that everything's fine and I'll merge.

I'm more than happy to publish a new release on PyPI after this, would you want to get the other PR through first?

jonded94 commented:
>would you want to get the other PR through first

Yes, surely, as we're using `mypy` to a broad degree in our entire codebase and this would be very helpful (I'm having to add `# type: ignore[import-untyped]` everywhere where I import this library right now).

Setting "sw mode" as an additional kwarg could be interesting for a later PR, after the release.
